subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/TaskExecution.java

```diff
@@ -384,6 +384,11 @@ public OverlappingOutputHandling getOverlappingOutputHandling() {
         return OverlappingOutputHandling.DETECT_OVERLAPS;
     }
 
+    @Override
+    public boolean shouldCleanupStaleOutputs() {
+        return context.getTaskExecutionMode().isTaskHistoryMaintained();
```

Comment:
Does this slightly change the behavior?

Previously we were checking `!task.getReasonNotToTrackState().isPresent()` in `CleanupStaleOutputsExecuter`. Should add this condition here as well?

Reply:
Good point. I first thought that this was a better way to express when we don't want to clean up the state, but indeed it changes the behavior slightly. Let's revert to the original for now.

Reply:
Actually, this is not a behavior change. If `task.getReasonNotToTrackState().isPresent()` is true, then the execution mode is `UNTRACKED`:

https://github.com/gradle/gradle/blob/e24b7f765af4e94ccfab7ff914ac7245df8f57f3/subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/DefaultTaskExecutionModeResolver.java#L39-L43

That's one of the two execution modes that returns false for `isTaskHistoryMaintained()`, the other being `NO_OUTPUTS`, in which case there are no outputs to clean up anyway.

I think this is a slightly better way to express the same behavior.

But this gave me an idea for an even better way to express the same behavior, hold on!

Reply:
Actually, no. Let's keep it this way for now. Later we should discuss how we should/can clean up stale outputs for transforms, if at all.